protocol/x/ratelimit/util/ibc.go

```diff
@@ -0,0 +1,191 @@
+package util
+
+// TODO(CORE-854): Improve attribution message.
+// This file re-uses similar utilities (with minor tweaking) Stride's IBC Rate Limit implementation:
+// https://github.com/Stride-Labs/stride/blob/121f2ac5d2e5f8e406f89999410a49ea4277a552/x/ratelimit
+
+import (
+	"encoding/json"
+	"fmt"
+	"math/big"
+
+	errorsmod "cosmossdk.io/errors"
+	"cosmossdk.io/log"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	ibctransfertypes "github.com/cosmos/ibc-go/v8/modules/apps/transfer/types"
+	channeltypes "github.com/cosmos/ibc-go/v8/modules/core/04-channel/types"
+	"github.com/dydxprotocol/v4-chain/protocol/x/ratelimit/types"
+)
+
+// Parse the denom from the Send Packet that will be used by the rate limit module
+// The denom that the rate limiter will use for a SEND packet depends on whether
+// it was a NATIVE token (e.g. ustrd, stuatom, etc.) or NON-NATIVE token (e.g. ibc/...)...
+//
+// We can identify if the token is native or not by parsing the trace denom from the packet
+// If the token is NATIVE, it will not have a prefix (e.g. ustrd),
+// and if it is NON-NATIVE, it will have a prefix (e.g. transfer/channel-2/uosmo)
+//
+// For NATIVE denoms, return as is (e.g. ustrd)
+// For NON-NATIVE denoms, take the ibc hash (e.g. hash "transfer/channel-2/usoms" into "ibc/...")
```

Comment:
IIUC, if the denom was IBC'd multiple hops, it has multiple prefixes. This is my interpretation from [this doc](https://tutorials.cosmos.network/academy/3-ibc/7-token-transfer.html)
> We can send assets (or their IBC voucher representation) in multiple hops across multiple chains. Every single time the path will be prepended with the port/channel-id/... prefix. When sending this IBC denom (having had multiple hops) back to its source chain, for every hop back one port/channel-id/... prefix will be taken off. This results in a return to the original denom if all the hops are reversed.

Does this remove all the prefixes for multi-hops? If not, does that cause an issue where the same tokens are considered different if the IBC path is different from capacity calculation perspective?

Reply:
Not exactly sure you mean by "Does this remove all the prefixes for multi-hops", but I believe the following is relevant to your question:


Consider the two different paths that `uusdc` travel to dYdX chain:

1. Noble -> dYdX Chain 
2. Noble -> Osmosis -> dYdX Chain

The resulting IBC denom on dYdX Chain is different, and therefore non-fungible, by design (more discussion [here](https://dydx-team.slack.com/archives/C063PG2P551/p1699377562092449?thread_ts=1699377113.925409&cid=C063PG2P551)). 
- Only path 1 uusdc will be accepted as collateral/received by users from trading. 
- Path 2 uusdc has NO significance on the dYdX chain. 

Therefore in the context of rate-limiting, only path 1 uusdc relevant for the ratelimit.